prov/shm/src/smr_ep.c

```diff
@@ -303,17 +308,46 @@ void smr_format_iov(struct smr_cmd *cmd, fi_addr_t peer_id,
 		    const struct iovec *iov, size_t count, size_t total_len,
 		    uint32_t op, uint64_t tag, uint64_t data, uint64_t op_flags,
 		    void *context, struct smr_region *smr,
-		    struct smr_resp *resp, struct smr_cmd *pend_cmd)
+		    struct smr_resp *resp, struct smr_tx_entry *pend)
 {
 	smr_generic_format(cmd, peer_id, op, tag, 0, 0, data, op_flags);
 	cmd->msg.hdr.op_src = smr_src_iov;
 	cmd->msg.hdr.src_data = (uintptr_t) ((char **) resp - (char **) smr);
 	cmd->msg.data.iov_count = count;
 	cmd->msg.hdr.size = total_len;
+	memcpy(cmd->msg.data.iov, iov, sizeof(*iov) * count);
+
+	smr_post_pend_resp(pend, cmd, context, iov, count, resp);
+}
+
+void smr_format_offset(struct smr_cmd *cmd, fi_addr_t peer_id,
+		       const struct iovec *iov, size_t count, size_t total_len,
+		       uint32_t op, uint64_t tag, uint8_t datatype,
+		       uint8_t atomic_op, uint64_t data, uint64_t op_flags,
+		       void *context, struct smr_region *smr,
+		       struct smr_region *peer_smr,
+		       struct smr_inject_buf **tx_bufs,
+		       uint64_t buf_count, struct smr_resp *resp,
+		       struct smr_tx_entry *pend)
+{
+	int i;
+
+	smr_generic_format(cmd, peer_id, op, tag, datatype, atomic_op, data, op_flags);
+	cmd->msg.hdr.op_src = smr_src_offset;
+	cmd->msg.hdr.src_data = (uintptr_t) ((char **) resp - (char **) smr);
+	cmd->msg.hdr.size = total_len;
 	cmd->msg.hdr.msg_id = (uint64_t) (uintptr_t) context;
+	cmd->msg.hdr.op_flags |= SMR_RMA_REQ;
+	cmd->msg.data.iov_count = count;
 	memcpy(cmd->msg.data.iov, iov, sizeof(*iov) * count);
 
-	smr_post_pend_resp(cmd, pend_cmd, resp);
+	cmd->msg.data.offset_count = buf_count;
+	for (i = 0; i < buf_count; i++)
+		cmd->msg.data.offset[i] = (uintptr_t)
+					((char **) tx_bufs[i] - (char **) peer_smr);
+
+	smr_post_pend_resp(pend, cmd, context, iov, count, resp);
```

Comment:
Please restrict what the format calls do.  They should only prepare/format a structure.  The posting of the send should come outside of this function, otherwise, it looks like a side effect and is not expected by the function name.

Reply:
I agree that we should restrict what the format calls do. 
I think `smr_post_pend_resp` is not doing `real` post. It just does a few assignments though, associating `resp` with `pend`, and `pend` with `cmd`.